txscript/consensus.go

```diff
@@ -25,12 +25,11 @@ const (
 // ExtractCoinbaseNullData ensures the passed script is a nulldata script as
 // required by the consensus rules for the coinbase output that is used to
 // ensure the coinbase has a unique hash and returns the data it pushes.
+//
+// NOTE: This function is only valid for version 0 scripts.  Since the function
+// does not accept a script version, the results are undefined for other script
+// versions.
 func ExtractCoinbaseNullData(pkScript []byte) ([]byte, error) {
```

Comment:
Is it possible to make this function valid for versions > 0?

Reply:
It's possible, but would require a v2 module since it requires the addition of a new version parameter which constitutes a major change according to semver.  However, it's important to note that no script versions greater than 0 are currently valid and this is simply calling out those existing semantics.

That said, part of introducing new script versions will necessarily involve ensuring the new `ScriptTokenizer` which this PR introduces and does take a script version, is able to properly parse the newly introduced script version(s).

To that end, this function this performs is now possible and trivial to implement via the new aforementioned `ScriptTokenizer` this PR introduces:

```Go
package main

import (
	"encoding/hex"
	"fmt"

	"github.com/decred/dcrd/txscript"
)

func main() {
	// This would oridinarily come from the tx output.
	const scriptVersion = 0

	// This is from main chain block 330233 and would ordinarily come from
	// somehwere else.
	pkScript, err := hex.DecodeString("6a0cf9090500f1ee035d7e430274")
	if err != nil {
		fmt.Println(err)
		return
	}

	// Note that this is ignoring the max allowed length which could be added
	// too, but for most applications, it won't matter hence I omitted it here.
	var data []byte
	if len(pkScript) > 1 && pkScript[0] == txscript.OP_RETURN {
		tokenizer := txscript.MakeScriptTokenizer(scriptVersion, pkScript[1:])
		if tokenizer.Next() && tokenizer.Done() &&
			tokenizer.Opcode() <= txscript.OP_PUSHDATA4 {

			data = tokenizer.Data()
		}
	}

	fmt.Printf("Coinbase data: %x", data)
}
```

Output:
```
Coinbase data: f9090500f1ee035d7e430274
```